bionic/deriver.py

```diff
@@ -140,38 +149,97 @@ def _get_ready_for_full_resolution(self):
             "core__versioning_policy"
         )
 
-        self._is_ready_for_full_resolution = True
+        self._bootstrap_is_complete = True
+
+    def _prevalidate_essential_dnodes(self):
+        """
+        Checks that all 'essential' descriptors can be computed.
+        """
 
-    def _get_ready_for_bootstrap_resolution(self):
-        if self._is_ready_for_bootstrap_resolution:
+        # Avoid doing pre-validation multiple times. (It's not that expensive since all
+        # the state is cached, but it's still O(number of descriptors), so we'll avoid
+        # it on principle.)
+        if self._essential_prevalidation_is_complete:
             return
 
-        # Generate the static key spaces and tasks for each descriptor.
-        self._key_spaces_by_dnode = {}
-        self._task_lists_by_dnode = {}
-        for name in self._flow_state.providers_by_name.keys():
-            dnode = DescriptorNode.from_descriptor(name)
-            self._populate_dnode_info(dnode)
+        self._prevalidate_dnodes(self._get_essential_dnodes())
+
+        self._essential_prevalidation_is_complete = True
+
+    def _prevalidate_dnodes(self, dnodes):
+        """
+        Identifies (and caches) all tasks required to compute a collection of dnodes.
+        Useful for surfacing any dependency errors ahead of time.
+        """
+
+        for dnode in dnodes:
+            dinfo = self._get_or_create_dinfo_for_dnode(dnode)
+            for task_key, task in dinfo.tasks_by_key.items():
+                self._get_or_create_task_state_for_key(task_key)
+
+    def _get_essential_dnodes(self):
+        """
+        Returns the list of descriptor nodes needed to compute all user-defined entities
```

Comment:
Same here, I'm not sure what an "essential" dnode is. Once we have descriptors, I'm not sure what all user-defined entities would mean either. How do you feel about base / basic dnodes?

Reply:
Yeah, "essential" is not a great name. "Base" or "basic" might be better, although I worry that they suggest the dnodes just have a simple structure. "Core" might be good if we weren't already using it elsewhere. Let's talk more about it tomorrow morning.